drgn_tools/multipath.py

```diff
@@ -0,0 +1,330 @@
+                pi.repeat_count.value_(),
+                " ",
+                load.value_(),
+            )
+        )
+    elif ps_type == "round-robin":
+        print(output_format.format(" ", pi.repeat_count.value_()))
+    elif ps_type == "queue-length":
+        print(
+            output_format.format(
+                " ",
+                pi.repeat_count.value_(),
+                " ",
+                pi.qlen.counter.value_(),
+            )
+        )
+    elif ps_type == "io-affinity":
+        print(
+            output_format.format(
+                " ",
+                pi.refcount.refs.counter.value_(),
+                " ",
+                pi.failed.value_(),
+            )
+        )
+    else:
+        print(output_format)
+
+
+def get_path_info(prog: Program, ps_type: str) -> Type:
+    ps_type_us = ps_type.replace("-", "_")
+    return type_lookup_conflict(
+        prog,
+        "struct path_info *",
+        f"dm_{ps_type_us}",  # module: e.g. dm_service_time
+        [f"dm-{ps_type}.c", f"dm-ps-{ps_type}.c"],  # e.g. dm-service-time.c
+    )
+
+
+def show_pgpaths(priority_group: Object) -> None:
+    """
+    Show pgpath data.
+
+    :param priority_group: `struct priority_group *`
+
+    :returns: None
+    """
+    path_num = 1
+    ps_type = priority_group.ps.type.name.string_().decode()
+    format_header = "{:10}{:11}{:12}{:10}"
+    print(
+        format_header.format("Path_No.", "Disk_Name", "DevName", "State"),
+        end="",
+    )
+    switch_header(ps_type)
+    path_info_type = get_path_info(priority_group.prog_, ps_type)
+    for pgpath in for_each_pgpath(priority_group):
+        bdev = pgpath.path.dev.bdev
+        disk_name = blkdev_name(bdev)
+        dev_name = pgpath.path.dev.name.string_().decode()
+        state = "active" if pgpath.is_active else "not active"
+        pi = cast(path_info_type, pgpath.path.pscontext)
+        format_val = "{:<3}{:<7}{:<2}{:<9}{:<1}{:<8}{:<3}{:<13}"
+        print(
+            format_val.format(
+                " ",
+                path_num,
+                " ",
+                disk_name,
+                " ",
+                dev_name,
+                " ",
+                state,
+            ),
+            end="",
+        )
+        switch_output(ps_type, pi)
+        path_num += 1
+
+
+def show_priority_groups(multipath: Object) -> None:
+    """
+    Show priority group prperties.
+
+    :param multipath: `struct multipath *`
+
+    :returns: None`
+    """
+    print_once = 1
+    for priority_group in for_each_priority_group(multipath):
+        path_selector_type = priority_group.ps.type.name.string_().decode()
+        if print_once:
+            print("Policy = %s" % path_selector_type)
+            print_once = 0
+        if path_selector_type == "historical-service-time":
+            print(
+                "\n\nPath selection policy %s not supported.\n\n"
+                % path_selector_type
+            )
+            return
```

Comment:
Why print  path selector type only once? What if different path group has different path selector type? I think all these dump path selector type logic should be merged into " show_pgpaths"
Also besides policy, can we simulate what "multipath -ll" on this? For example, here is the output for round-robin policy.
```
policy='round-robin 0' prio=30 status=active
```

Reply:
prio and status is maintained at application level I didn't any corresponding field in kernel. 

Kernel only has one field which is for individual path status.  it can be either active or failed.  eg.

eg. I failed one path using multipathd -k (interactive mode).

[root@shmsingh-multipath ~]# multipath -ll
mpathj (360014052e39672a288a48f1a11dfb817) dm-2 LIO-ORG,blockOL9
size=100G features='0' hwhandler='1 alua' wp=rw
|-+- policy='round-robin 0' prio=50 status=enabled
| `- 6:0:0:0 sde 8:64 failed ready running      <<== path showing as failed.
`-+- policy='round-robin 0' prio=50 status=active
  `- 7:0:0:0 sdd 8:48 active ready running


drgn script output:

[root@shmsingh-multipath drgn-tools]# python3 -m drgn_tools.corelens /proc/kcore  -M multipath
====== MODULE multipath ======
--------------- Device [1] ---------------
mpathj (360014052e39672a288a48f1a11dfb817) dm-2
struct multipath * (0xffff98a389467800)
Flags = 0
Flags = 0x0
Current path = 8:48
hw_handler_name = alua
gendisk = 0xffff98a387ef4200
bdev = 0xffff98a380419c00
Permission = rw
Size = 107374182400
Policy = round-robin
Path_No.  Disk_Name  DevName     State        Repeat_Count
   1        sde       8:64       not active   <<=== showing as not active.     1
Path_No.  Disk_Name  DevName     State        Repeat_Count
   1        sdd       8:48       active

Reply:
I am looking in multipath application code if it can be implemented in drgn.

Reply:
I see, we can have drgn show the information if they were only maintained in the userspace. Let skip them. Just have policy shown for each priority group.

Reply:
I will update the code to print policy for each priority group.